--- a/target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh
+++ b/target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh
@@ -110,7 +110,6 @@ platform_do_upgrade() {
 	bananapi,bpi-r4-lite|\
 	bazis,ax3000wm|\
 	cmcc,a10-ubootmod|\
-	cmcc,rax3000m|\
 	comfast,cf-wr632ax-ubootmod|\
 	creatlentem,clt-r30b1-ubi|\
 	cudy,m3000-v1-ubootmod|\
@@ -122,12 +121,7 @@ platform_do_upgrade() {
 	cudy,wr3000h-v1-ubootmod|\
 	cudy,wr3000p-v1-ubootmod|\
 	gatonetworks,gdsp|\
-	globitel,bt-r320|\
-	h3c,magic-nx30-pro|\
 	imou,hx21|\
-	jcg,q30-pro|\
-	jdcloud,re-cp-03|\
-	konka,komi-a31|\
 	mediatek,mt7981-rfb|\
 	mediatek,mt7988a-rfb|\
 	mercusys,mr90x-v1-ubi|\
@@ -136,20 +130,12 @@ platform_do_upgrade() {
 	netis,nx30v2|\
 	netis,nx31|\
 	netis,nx32u|\
-	nokia,ea0326gmp|\
 	openwrt,one|\
-	netcore,n60|\
-	netcore,n60-pro|\
-	qihoo,360t7|\
 	qihoo,360t7-ubi|\
 	routerich,ax3000-ubootmod|\
 	routerich,be7200|\
 	snr,snr-cpe-ax2|\
 	teralink,tl3020-256mb|\
-	tplink,tl-xdr4288|\
-	tplink,tl-xdr6086|\
-	tplink,tl-xdr6088|\
-	tplink,tl-xtr8488|\
 	wavlink,wl-wnt100x3-ubootmod|\
 	xiaomi,mi-router-ax3000t-ubootmod|\
 	xiaomi,redmi-router-ax6000-ubootmod|\
@@ -163,6 +149,9 @@ platform_do_upgrade() {
 	acer,vero-w6m|\
 	airpi,ap3000m|\
 	arcadyan,mozart|\
+	cmcc,rax3000m-emmc|\
+	cmcc,xr30-emmc|\
+	sl,3000-emmc|\
 	glinet,gl-mt2500|\
 	glinet,gl-mt2500-airoha|\
 	glinet,gl-mt6000|\
@@ -324,7 +313,6 @@ platform_check_image() {
 	bananapi,bpi-r4-lite|\
 	bazis,ax3000wm|\
 	cmcc,a10-ubootmod|\
-	cmcc,rax3000m|\
 	comfast,cf-wr632ax-ubootmod|\
 	creatlentem,clt-r30b1-ubi|\
 	cudy,m3000-v1-ubootmod|\
@@ -336,28 +324,16 @@ platform_check_image() {
 	cudy,wr3000h-v1-ubootmod|\
 	cudy,wr3000p-v1-ubootmod|\
 	gatonetworks,gdsp|\
-	globitel,bt-r320|\
-	h3c,magic-nx30-pro|\
-	jcg,q30-pro|\
-	jdcloud,re-cp-03|\
-	konka,komi-a31|\
 	mediatek,mt7981-rfb|\
 	mediatek,mt7988a-rfb|\
 	mercusys,mr90x-v1-ubi|\
-	nokia,ea0326gmp|\
 	netis,eap930-v1|\
 	netis,n6-v2|\
 	netis,nx32u|\
 	openwrt,one|\
-	netcore,n60|\
-	qihoo,360t7|\
 	qihoo,360t7-ubi|\
 	routerich,ax3000-ubootmod|\
 	teralink,tl3020-256mb|\
-	tplink,tl-xdr4288|\
-	tplink,tl-xdr6086|\
-	tplink,tl-xdr6088|\
-	tplink,tl-xtr8488|\
 	wavlink,wl-wnt100x3-ubootmod|\
 	xiaomi,mi-router-ax3000t-ubootmod|\
 	xiaomi,redmi-router-ax6000-ubootmod|\
@@ -409,6 +385,9 @@ platform_copy_config() {
 	acer,vero-w6m|\
 	airpi,ap3000m|\
 	arcadyan,mozart|\
+	cmcc,rax3000m-emmc|\
+	cmcc,xr30-emmc|\
+	sl,3000-emmc|\
 	glinet,gl-mt2500|\
 	glinet,gl-mt2500-airoha|\
 	glinet,gl-mt6000|\
@@ -452,10 +431,19 @@ platform_pre_upgrade() {
 	jiorouter,ax6000-jidu6101)
 		jiorouter_initial_setup
 		;;
-	xiaomi,mi-router-ax3000t|\
-	xiaomi,mi-router-wr30u-stock|\
-	xiaomi,redmi-router-ax6000-stock)
-		xiaomi_initial_setup
+	cmcc,mr3000d-ciq-256m)
+		tenbay_dualboot_fixup
 		;;
 	esac
 }
+
+tenbay_dualboot_fixup() {
+	[ "$(rootfs_type)" = "tmpfs" ] || return 0
+
+	if ! fw_printenv -n boot_from &>/dev/null; then
+		echo "unable to read uboot-env"
+		return 1
+	fi
+
+	fw_setenv boot_from ubi
+}
